libutils/glob_lib.c

```diff
@@ -0,0 +1,446 @@
+
+        switch (ch)
+        {
+        case '*':
+            BufferAppendString(buf, ".*");
+            break;
+
+        case '?':
+            BufferAppendString(buf, ".");
+            break;
+
+        case '[':
+            i = TranslateBracket(pattern, n, i, buf);
+            break;
+
+        default:
+            if (strchr(special_chars, ch) != NULL)
+            {
+                BufferAppendF(buf, "\\%c", ch);
+            }
+            else
+            {
+                BufferAppendChar(buf, ch);
+            }
+            break;
+        }
+    }
+
+    char *const res = StringFormat("(?s:%s)\\Z", BufferData(buf));
+    BufferDestroy(buf);
+    return res;
+}
+
+bool GlobMatch(const char *const _pattern, const char *const _filename)
+{
+    assert(_pattern != NULL);
+    assert(_filename != NULL);
+
+#ifdef _WIN32
+    // Case-normalize pattern and filename on Windows
+    char *pattern = xstrdup(_pattern);
+    ToLowerStrInplace(pattern);
+    char *filename = xstrdup(_filename);
+    ToLowerStrInplace(filename);
+#else  // _WIN32
+    const char *pattern = _pattern;
+    const char *filename = _filename;
+#endif // _WIN32
+
+    StringSet *const expanded = StringSetNew();
+    ExpandBraces(pattern, expanded);
+
+#ifdef _WIN32
+    free(pattern);
+    pattern = NULL;
+#endif // _WIN32
+
+    bool any_matches = false;
+    StringSetIterator iter = StringSetIteratorInit(expanded);
+    for (pattern = StringSetIteratorNext(&iter); pattern != NULL;
+         pattern = StringSetIteratorNext(&iter))
+    {
+        char *const regex = TranslateGlob(pattern);
+        any_matches |= StringMatchFull(regex, filename);
+        free(regex);
+    }
+    StringSetDestroy(expanded);
+
+#ifdef _WIN32
+    free(filename);
+    filename = NULL;
+#endif // _WIN32
+
+    return any_matches;
+}
+
+/**
+ * Strategically walk through filesystem looking for files matching pattern.
+ */
+static void WalkPattern(
+    const char *const root,
+    const Seq *const components,
+    const size_t index,
+    Seq *const matches)
+{
+    assert(components != NULL);
+    assert(matches != NULL);
+
+    const size_t num_components = SeqLength(components);
+    if (index >= num_components)
+    {
+        // We have found a match!
+        char *const match = xstrdup(root);
+        SeqAppend(matches, match);
+        return;
+    }
+
+    char *const next = (root == NULL) ? xstrdup(SeqAt(components, index))
+                                      : StringFormat(
+                                          "%s" FILE_SEPARATOR_STR "%s",
+                                          root,
+                                          (char *) SeqAt(components, index));
+
+    if (index == 0)
+    {
+        if (strlen(next) == 2 && isalnum(next[0]) && next[1] == ':')
+        {
+            WalkPattern(next, components, index + 1, matches);
+            free(next);
+            return;
+        }
+        if (StringEqual(next, ""))
```

Comment:
Since you're using `StringEqual` and other `string_lib` functions, might as well use `SafeStringLength()` too.

Reply:
I would rather crash and discover it early if `next == NULL`

Reply:
You're no longer using `strlen()` in this function, however note that relying on `strlen()`, `strcmp()`, etc. to crash is a bad idea. If you're lucky, it crashes, if not, you could have a security vulnerability :)

Reply:
The only difference is that `SafeStringLength()` checks for `NULL`. Accessing memory address `0x0` will always cause the program to crash.